code/go/0chain.net/smartcontract/storagesc/challenge.go

```diff
@@ -722,39 +827,77 @@ func (sc *StorageSmartContract) addChallenge(alloc *StorageAllocation,
 
 	var storageChallenge StorageChallenge
 	storageChallenge.ID = challengeID
-	storageChallenge.Validators = selectedValidators
-	storageChallenge.Blobber = selectedBlobberObj
+	//storageChallenge.Validators = selectedValidators
+	storageChallenge.TotalValidators = len(selectedValidators)
+	//storageChallenge.Blobber = blobber
+	storageChallenge.BlobberID = blobberID
 	storageChallenge.RandomNumber = challengeSeed
 	storageChallenge.AllocationID = alloc.ID
-
 	storageChallenge.AllocationRoot = blobberAllocation.AllocationRoot
+	storageChallenge.Created = creationDate
+
+	blobberChallengeObj, err := sc.getBlobberChallenge(blobberID, balances)
+	if err != nil {
+		if err == util.ErrValueNotPresent {
+			blobberChallengeObj = &BlobberChallenge{}
+			blobberChallengeObj.BlobberID = blobberID
+		} else {
+			return "", common.NewError("add_challenge",
+				"error fetching blobber challenge: "+err.Error())
+		}
+	}
 
-	blobberChallengeObj := &BlobberChallenge{}
-	blobberChallengeObj.BlobberID = storageChallenge.Blobber.ID
+	addedChallenge := blobberChallengeObj.addChallenge(&storageChallenge)
+	if !addedChallenge {
+		challengeBytes, err := json.Marshal(storageChallenge)
+		return string(challengeBytes), err
+	}
 
-	blobberChallengeBytes, _ := balances.GetTrieNode(blobberChallengeObj.GetKey(sc.ID))
-	if blobberChallengeBytes != nil {
-		err = blobberChallengeObj.Decode(blobberChallengeBytes.Encode())
-		if err != nil {
-			return "", common.NewError("blobber_challenge_decode_error",
-				"Error decoding the blobber challenge")
+	allocChallengeObj, err := sc.getAllocationChallenge(alloc.ID, balances)
+	if err != nil {
+		if err == util.ErrValueNotPresent {
+			allocChallengeObj = &AllocationChallenge{}
+			allocChallengeObj.AllocationID = alloc.ID
+		} else {
+			return "", common.NewError("add_challenge",
+				"error fetching allocation challenge: "+err.Error())
 		}
 	}
 
-	storageChallenge.Created = creationDate
-	addedChallege := blobberChallengeObj.addChallenge(&storageChallenge)
-	if !addedChallege {
+	addedAllocChallenge := allocChallengeObj.addChallenge(&storageChallenge)
+	if !addedAllocChallenge {
 		challengeBytes, err := json.Marshal(storageChallenge)
 		return string(challengeBytes), err
 	}
 
-	balances.InsertTrieNode(blobberChallengeObj.GetKey(sc.ID), blobberChallengeObj)
+	_, err = balances.InsertTrieNode(allocChallengeObj.GetKey(sc.ID), allocChallengeObj)
+	if err != nil {
+		return "", common.NewError("add_challenge",
+			"error storing alloc challenge: "+err.Error())
+	}
+
+	_, err = balances.InsertTrieNode(storageChallenge.GetKey(sc.ID), &storageChallenge)
```

Comment:
The storageChallenge object should be emitted to the event database and stored in the Challenge table, not stored in the MPT. For example see https://github.com/0chain/0chain/pull/679 
[balances.EmitEvent(event.TypeStats, event.TagAddChallenge, challengeID, string(data))](https://github.com/0chain/0chain/pull/679/files?show-viewed-files=true&file-filters%5B%5D=#diff-820fe8e15b4ee1989877f5d353270c462b240ff6a69049e62161ae03d27727eeR809)

This way we can remove everything from StorageChallenge except what we actually need for smart contracts. Ideally, it can be replaced altogether with summary stats.